Objects/dictobject.c

```diff
@@ -7222,37 +7222,39 @@ _PyObjectDict_SetItem(PyTypeObject *tp, PyObject **dictptr,
     PyInterpreterState *interp = _PyInterpreterState_GET();
 
     assert(dictptr != NULL);
-    if ((tp->tp_flags & Py_TPFLAGS_HEAPTYPE) && (cached = CACHED_KEYS(tp))) {
-        assert(dictptr != NULL);
+    dict = *dictptr;
```

Comment:
We need to use atomics to load from the `dictptr` or there will be a data race between the read here and the write within the critical section.